android/app/src/test/java/com/threed/supply/LoginLogicTest.kt

```diff
@@ -0,0 +1,165 @@
+import com.threed.supply.features.login.domain.LoginEffect.*
+import com.threed.supply.features.login.domain.LoginEvent
+import com.threed.supply.features.login.domain.LoginEvent.*
+import com.threed.supply.features.login.domain.LoginModel
+import com.threed.supply.features.login.domain.update
+import org.junit.Before
+import org.junit.Test
+
+class LoginLogicTest {
+
+    private lateinit var updateSpec: UpdateSpec<LoginModel, LoginEvent, LoginEffect>
+
+    @Before
+    fun setUp() {
+        updateSpec = UpdateSpec(::update)
+    }
+
+    @Test
+    fun updateEmail_whenEmailChanged() {
+        val model = LoginModel()
+        updateSpec
+            .given(model)
+            .whenEvent(EmailChanged(email = "test"))
+            .then(assertThatNext(hasModel(model.copy(email = "test"))))
+    }
+
+    @Test
+    fun updatePassword_whenPasswordChanged() {
+        val model = LoginModel()
+        updateSpec
+            .given(model)
+            .whenEvent(PasswordChanged(password = "1234"))
+            .then(assertThatNext(hasModel(model.copy(password = "1234"))))
+    }
+
+    @Test
+    fun enableLogin_whenEmailChanged_andValidEmailAndPassword() {
+        val model = LoginModel(password = "12345678")
+        updateSpec
+            .given(model)
+            .whenEvent(EmailChanged(email = "test@gmail.com"))
+            .then(
+                assertThatNext(
+                    hasModel(model.copy(email = "test@gmail.com", loginEnabled = true))
+                )
+            )
+    }
+
+    @Test
+    fun enableLogin_whenPasswordChanged_andValidEmailAndPassword() {
+        val model = LoginModel(email = "test@gmail.com")
+        updateSpec
+            .given(model)
+            .whenEvent(PasswordChanged(password = "12345678"))
+            .then(
+                assertThatNext(
+                    hasModel(model.copy(password = "12345678", loginEnabled = true))
+                )
+            )
+    }
+
+    @Test
+    fun disableLogin_whenInvalidEmail() {
+        val model = LoginModel(email = "test@gmail.com", loginEnabled = true)
+        updateSpec
+            .given(model)
+            .whenEvent(EmailChanged(email = "test@gmail.com"))
+            .then(
+                assertThatNext(
+                    hasModel(model.copy(email = "test@gmail.com", loginEnabled = false))
+                )
+            )
+    }
+
+    @Test
+    fun disableLogin_whenInvalidPassword() {
+        val model = LoginModel(email = "test@gmail.com", loginEnabled = true)
+        updateSpec
+            .given(model)
+            .whenEvent(PasswordChanged(password = "1234567"))
+            .then(
+                assertThatNext(
+                    hasModel(model.copy(password = "1234567", loginEnabled = false))
+                )
+            )
+    }
+
+    @Test
+    fun login_whenLoginRequested() {
+        val model = LoginModel(
+            email = "test@gmail.com",
+            password = "1234",
```

Comment:
The login shouldn't be successful with an invalid password. I assume this doesn't fail as there is no validation in the loop yet, right?